### ISSUE TITLE
Tagged unions don't differentiate types with Literal Boolean values

### ISSUE THREAD

**Issue by w0rp** [microsoft/pyright#1603]
**Describe the bug**

If you use `Literal[True]` and `Literal[False]` to test items in a tagged union, Pyright doesn't infer the correct types.

**To Reproduce**

Check the example code using the latest version of Pyright in a Python 3.6 codebase with `typing_extensions`.

**Expected behavior**

You should be able to infer the type from a `Union` with literal Boolean values.

**Screenshots or Code**

See the following example code.

```python
from typing import Union

from typing_extensions import Literal

class Foo:
    tag: Literal[False] = False

class Bar:
    tag: Literal[True] = True

Either = Union[Foo, Bar]

def f(obj: Either) -> None:
    if obj.tag:  # obj.tag here is `bool`
        obj  # The type here should be `Bar`, but is `Either`
```

**VS Code extension or command-line**

I'm running Pyright via the LSP server through Vim and ALE.

**Additional context**

I'm using a Python 3.6 codebase with typing extensions. I'm not sure if this problem is specific to the use of typing_extensions, or if it also happens with the standard typing module in newer Python versions.

If this can be fixed, it is particularly useful for inferring better types when checking between authenticated and anonymous user types in Django.

**Comment by w0rp** [comment]
I can use `== False`, but it is something the type checking system should be able to figure out. `== False` violates PEP8 rule E712, so in order to get the tagged union to work properly you'd have to ignore a PEP8 rule with a comment. It will look strange in your code.

I don't think you should just close issues to keep an issue count low. Having 100 open issues doesn't harm a project. Using Boolean values as tags is important for Django, see here: https://docs.djangoproject.com/en/3.1/ref/contrib/auth/#django.contrib.auth.models.User.is_authenticated As more Django users adopt Pyright, they'll be confused about why the type system can't figure out `user.is_authenticated and user.authenticated_user_only_property == 123`, and they'll need to learn to go over their code and change it to `user.is_authenticated == True and user.authenticated_user_only_property == 123 # noqa: E712 `
